### DISCUSSION TITLE
Todo List

### DISCUSSION THREAD

**Discussion by tabreturn** [tabreturn/thonny-py5mode#17] [Ideas]
## Prioritize 

- [ ] Support for portable Thonny? Provide pre-packaged portable thonny+py5 releases, with a build script for portable mode (using villares' approach below)
- [ ] See issues: https://github.com/tabreturn/thonny-py5mode/issues

## Lower priority

- [ ] Restorable Thonny settings -- restore applied py5 config (themes, etc.) on deactivate
- [ ] Set interpreter (in config) to Thonny default (user might have selected another)
- [ ] Configure GitHub Actions to publish package? (and other useful actions, too?)
- [ ] pyp5js-powered exporter (to export thonny-py5mode sketches for javascript/web)
- [ ] Kill existing sketches when running a new one

## To Consider

- Thonny 4 includes an *Automatically propose completions while typing* option (under **Tools > Options > Editor**). It's pretty neat. Should this be activated by **py5 > Apply recommended py5 settings**?

## Done

- [x] Integrate alpha version py5 (and JDK) installer
- [x] Support for non-portable/installed Thonny
- [x] Imported mode auto-completion
- [x] Processing inspired theme for UI
- [x] Processing inspired theme for syntax
- [x] Imported mode Syntax highlighting
- [x] Pop up built-in save dialog for new unsaved files
- [x] Update instructions/readme
- [x] Add 'about' item under py5 menu (for plugin, py5, etc. version info and credits)?
- [x] Add Processing.py mode (that employs py5_tools > translators)?
- [x] Conceal full run_sketch.py run command in Thonny shell?
- [x] Add clear console to recommended py5 settings? (https://github.com/thonny/thonny/issues/1681) 
- [x] Remember sketch window position (https://github.com/hx2A/py5generator/issues/60)
- [x] Add py5 menu links to resources (ixora.io reference and cheatsheet)
- [x] Add py5 menu color selector
- [x] Display download/installation progress in Thonny (so Thonny doesn't appear frozen while downloading JDK)

**Comment by villares** [comment]
Hi! So... I've been thinking about the portable Thonny support... I need your opinions!

We want it to work "out of the box" right?
As it is now, either the java-home-in-thonny-options will be empty or wrong in the first run. The person would have to activate the menu for it to be set correctly *and also* one would need to restart Thonny. Doesn't seem easy enough.

I have uploaded it to dropbox, [thonny-4b3-py5-pymunk-win.zip](https://www.dropbox.com/s/3ue4cx3yf372teg/thonny-4b3-py5-pymunk-win.zip?dl=0), but here is the thing:
- I disabled the `install_jdk()` call when the menu is toggled (unnecessary, jdk-17 is installed!)
- I kept empty the environment variables in Thonny options (no JAVA_HOME)
- I made a `Thonny4.cmd` launcher that always sets  JAVA_HOME before starting `thonny.exe`
- ~~I made `thonny.exe` a hidden file (I'm unsure is this is a good idea).~~
- it seems to work!

I just commented out the line with `install_jdk()` we could do it like this when preparing portables, or, if you don't think it is overkill, whe could check for the presence of `portable_thonny.ini` before calling it.

**Reply by tabreturn** [reply]
Yup -- I agree that working right 'out of the box' is the way to go! I tested your download, and it seems to work fine on my Windows machine :)

I suppose I could add something that checks for the JDK directory in the portable package (as well as the usual location), so the `install_jdk()` call is conditional?

Technically, you could use the .cmd file to run a Python script (using the python.exe) to pull off something more elaborate (if you want to set the Java path with Python, edit/amend any configuration files, etc.)

I think we should include these portable versions in the GitHub releases page of the thonny-py5mode repo? I'm just wondering if we should look at a 'build' solution for generating releases (that automatically downloads and bundles Thonny and the relevant packages and applies the necessary configuration and tweaks). We should probably delete any .pyc files and \_\_pycache\_\_ directories in the process (to speed up the extraction process, if only a bit)?

**Reply by villares** [reply]
> Technically, you could use the .cmd file to run a Python script (using the python.exe) to pull off something more elaborate (if you want to set the Java path with Python, edit/amend any configuration files, etc.)

Yeah, but I don't think we need it at this point. Maybe it is best for a portable not to change anything permanently in the host computer?

> I think we should include these portable versions in the GitHub releases page of the thonny-py5mode repo? I'm just wondering if we should look at a 'build' solution for generating releases (that automatically downloads and bundles Thonny and the relevant packages and applies the necessary configuration and tweaks).

This would be super cool. I'm afraid my GitHub account doesn't like big files though, you might need to check that out too.

> We should probably delete any .pyc files and __pycache__ directories in the process (to speed up the extraction process, if only a bit)?

Oh, I completely missed this. Brilliant.

**Reply by tabreturn** [reply]
> This would be super cool. I'm afraid my GitHub account doesn't like big files though, you might need to check that out too.

The file size limit is 2 GB for release files, I think. I uploaded your archive to test, which uploaded fine --
https://github.com/tabreturn/thonny-py5mode/releases/tag/v0.3.3a0

**Comment by villares** [comment]
Cheers @tabreturn! How are things going in your new home?

I had a thought about the plug-in you might be interested to hear...
Could we, and should we, decouple the "initial JDK install" from the menu toggle named "imported mode for py5"?

My thinking was, would it be more clear for users to have the plugin check JDK status on load and display either the working menus or a different menu named "install JDK for py5",  "thonny-py5mode first-run" or something else. 

I can't remember much about my portable plug-in hacking  (the code present in my portable windows zip-package), it had to do with searching for JDK and updating the Thonny options environment variables field so that  "JAVA_HOME"  would be always updated even if you moved Thonny's folder. But I could could go through that messy code with you again if you want.

**Comment by villares** [comment]
About the "lower priority" 
> pyp5js-powered exporter (to export thonny-py5mode sketches for javascript/web)

I have made this pyp5js hack with most of py5 names: http://abav.lugaralgum.com/pyp5js/py5mode
It also adds a `background(200)` to `size()` and some other small `color_mode()` adaptations to make it look more like py5.

**Reply by tabreturn** [reply]
Thanks, @villares!

Sorry about the silence/inactivity. My last job left no room for working on this. I've just taken on a new position and plan to pick up work on py5 stuff again.

**Reply by villares** [reply]
Cheers, Tristan! Good to hear from you!

Something a bit more important for the "to do list":

Is the plug-in install causing the install of py5 0.8X? Could you check if it can install 0.91+?

I'm terrible with package stuff but I have the impression the plug-in requirements is pined to the older version and this is causing all sorts of trouble in class...

**Reply by villares** [reply]
Nota a priority, but a small friction students have: you can't run a sketch if the previous one is still running. Could we stop the running sketch when the run command is pressed? This would allow for a  quicker feedback loop I guess.

**Reply by hx2A** [reply]
> Could we stop the running sketch when the run command is pressed?

That would require killing the process. @tabreturn, would the plugin need a process ID number? Or do you get one from the plugin's method for launching the process?

> Is the plug-in install causing the install of py5 0.8X? Could you check if it can install 0.91+?

Every time I look at py5's charts on [pepy.tech](https://www.pepy.tech/projects/py5?versions=0.9.1a1&versions=0.8.0a2) I see people are installing an older version. For some reason, 0.8.0a2 gets more installs than the other 0.8.* packages.

Good to hear from you, Tristan! Glad you now have more time for stuff outside of work!

**Reply by tabreturn** [reply]
@villares I always use the stop button in the Thonny interface to stop the sketch. I tried something that kills any existing sketch, but I didn't get far. I'll look at this again in time; I cannot recall exactly where I got stuck last time. @hx2A Regarding process ID vs. launcher method, I'll have a closer look.

Bumping packages is probably one of the first things I'll tackle. That, and ensuring compatibility with 4.1.x

**Reply by hx2A** [reply]
@tabreturn, I seem to remember you once pinned the version number because of a py5 release that negatively affected Thonny users. Perhaps that was the right thing to do then, but now that code is fairly stable and doesn't change much. Now you probably could set the version number with a `>=` requirement and get good results.

On the subject of py5 releases, I'd like to do a release in October. I have coded some nice integrations with shapely and trimesh but they require a Processing bug fix I made a PR for. I want to at least know if my PR will be accepted or not before moving forward with those new py5 features. If not, I'll make some design changes to compensate.

I'm also moving forward with some new color features for py5. That will be done soon. The shapely and trimesh stuff may wait for some other future release.
